src/Multipart/AbstractUploader.php

```diff
@@ -0,0 +1,455 @@
+
+/**
+ * Encapsulates the execution of a multipart upload to S3 or Glacier.
+ */
+abstract class AbstractUploader
+{
+    /** @var Client Client used for the upload. */
+    protected $client;
+
+    /** @var array Configuration used to perform the upload. */
+    protected $config;
+
+    /** @var array Service-specific information about the upload workflow. */
+    protected $info = [];
+
+    /** @var PromiseInterface Promise that represents the multipart upload. */
+    protected $promise;
+
+    /** @var Stream Source of the data to be uploaded. */
+    protected $source;
+
+    /** @var UploadState State used to manage the upload. */
+    protected $state;
+
+    /**
+     * Construct a new transfer object
+     *
+     * @param Client $client
+     * @param mixed  $source
+     * @param array  $config
+     */
+    public function __construct(Client $client, $source, array $config = [])
+    {
+        $this->loadUploadWorkflowInfo();
+        $this->config = $config + $this->info['defaults'] + [
+            'part_size'   => null,
+            'upload_id'   => null,
+            'state'       => null,
+            'concurrency' => 3,
+            'before_part' => null,
+            'initiate'    => [],
+            'upload'      => [],
+            'complete'    => [],
+            'abort'       => [],
+        ];
+        $this->client = $client;
+        $this->prepareSource($source);
+        $this->prepareState();
+    }
+
+    /**
+     * Returns the current state of the upload
+     *
+     * @return UploadState
+     */
+    public function getState()
+    {
+        if ($this->state instanceof Promise\PromiseInterface) {
+            $this->state = $this->state->wait();
+        }
+
+        return $this->state;
+    }
+
+    /**
+     * Upload the source using multipart upload operations.
+     *
+     * @return Result The result of the CompleteMultipartUpload operation.
+     * @throws \LogicException if the upload is already complete or aborted.
+     * @throws MultipartUploadException if an upload operation fails.
+     */
+    public function upload()
+    {
+        return $this->uploadAsync()->wait();
+    }
+
+    /**
+     * Abort the multipart upload.
+     *
+     * @return Result
+     * @throws \LogicException if the upload has not been initiated yet.
+     * @throws MultipartUploadException if the abort operation fails.
+     */
+    public function abort()
+    {
+        return $this->abortAsync()->wait();
+    }
+
+    /**
+     * Upload the source asynchronously using multipart upload operations.
+     *
+     * @return PromiseInterface
+     */
+    public function uploadAsync()
+    {
+        if ($this->promise) {
+            return $this->promise;
+        }
+
+        return $this->promise = Promise\coroutine(function () {
+            // Determine the state of the upload
+            if ($this->state instanceof Promise\PromiseInterface) {
+                $this->state = (yield $this->state);
```

Comment:
You can yield regular values and promises. I suggest yielding regardless of it's a promise or not (but I also suggest always making it a promise).